### DISCUSSION TITLE
Save torrent magnet to blackhole folder

### DISCUSSION THREAD

**Discussion by felixbrucker** [pymedusa/Medusa#7407] [Ideas & Feature requests]
I'm currently using sickchill with my own patches to add torrent magnet links to the torrent backhole folder as `xyz.magnet` which is basically a text file with the magnet link inside.

Could this feature be added to medusa as well?

The (pretty old) changes i made for reference: https://github.com/felixbrucker/SickRage/compare/master...felixbrucker:felixbrucker/magnet-blackhole

**Comment by p0psicles** [comment]
I don't see why not. @medariox?

**Comment by medariox** [comment]
I don't see the point of it. What application uses these .magnet files?

**Comment by felixbrucker** [comment]
The problem i solved with this is the following:
Torrent providers might provide .torrent files and/or magnet files. In the event that the torrent provider only provides a magnet link the blackhole torrent downloader needs a torrent tho, so a "converter" is used to "convert" the magnet to a torrent file, which doesnt work reliably/at all in my experience.
As for the torrent downloader im using, im using real-debrid (https://real-debrid.com/) which downloads .torrents or magnets and provides a http download of the finished torrent. I have written a small app that watches a folder for .torrent or .magnet files, adds them to the service, downloads the finished files, unpacks them and moves them into their destination where sickchill can find them.

If you want to support that service directly that would be an alternative of course, but very likely more work.

**Comment by medariox** [comment]
The problem I have with the 1. approach is that it will only serve it's purpose with your exact workflow and scripts. It will be code added for thousands of users that will literally benefit just one person.

I think that a proper implementation, even if initially more work intensive, would allow more users to benefit from it long term and result in a real alternative to the current methods.

**Comment by gersilex** [comment]
Using the exact same workflow with different software. I think adding a feature to "just save the magnet link in a file" instead of failing sounds reasonable. I see how not everybody has this use-case, but that's probably the case for many features.

Just wanted bump this to make clear that there really is interest in. I'm using sonarr just for this single feature and I really don't like it but have no choice.

**Comment by hurricane-dorian** [comment]
Since there is a bunch of debrider providers now, support for all of them can be provided via this feature and it would be very useful even if specific support for real-debrid is added.

**Comment by morpheasgr** [comment]
+1 for debrider service/cached torrent support. You'd have most of the Kodi community jumping in. Especially since Real-Debrid, Premiumize's and AllDebrid's api also enables directly streaming or adding to the user's 'cloud' from their cache instead of having to download.

**Comment by p0psicles** [comment]
Anyone here that wants to take the time to explain possible solutions on this?
Please contact me on our discord.

**Comment by gersilex** [comment]
@p0psicles and me had a great talk in Discord and we were able to furthe drill down into the why's and what's of this topic. We concluded the following status quo:

- Torrent providers provide `torrent` files or `magnet:` URIs
- Some providers only provide (much much smaller) `magnet:` URIs
- Medusa has a feature to look up the magnet hash in [public cache databases](https://github.com/pymedusa/Medusa/blob/b352bb6924afcdfafce176a540d53ce405ca1312/medusa/providers/torrent/torrent_provider.py#L36)
- If all of those caches are unavailable, the snatch fails and no `torrent` file is saved.
- Some download services can make use of the `magnet:` URIs themselves and download the files anyway
- A `magnet:` URI can be derived (hashed) from the content of a `torrent` file.

Above facts show that a `torrent` file is the most useful as it contains all data necessary for a download service provider or torrent client. However, it is not strictly required and a `magnet:` URI is enough to download those files. 

We would like to propose the following behavior (_Functionality already present in Medusa is checked_):

- [x] Download torrent data from the provider
- [x] Save it as a `.torrent` file
- [x] If the provider answers with a `magnet:` URI: Find and download a `.torrent` file from the torrent cache database
- [ ] If the lookup fails (because nothing was found or no database is reachable) AND the user enabled this feature, save the `magnet:` URI into a `.magnet` file
- [x] If the feature is not enabled, fail with a message in the log files.

So we basically would add one step in-between that does not break the current behavior but gives users of the Blackhole more freedom what to do with the snatched data. Our friends over at [Sonarr do this](https://github.com/Sonarr/Sonarr/blob/25baf7bb45d92bdeb51063d5a14f8d1d9bc02253/src/NzbDrone.Core/Download/Clients/Blackhole/TorrentBlackhole.cs#L60) for a while now and it deemed very useful to the users, me included.

Please let us know what you think!

**Reply by felixbrucker** [reply]
Works for me and doesn't break my existing flow, except for finally switching to medusa of course 👍
